node/network/collator-protocol/src/peer_slots.rs

```diff
@@ -0,0 +1,639 @@
+			);
+		}
+	}
+
+	pub(crate) fn remove_outgoing(
+		&mut self,
+		old_relay_parents: impl IntoIterator<Item = Hash>,
+	) {
+		for old_relay_parent in old_relay_parents {
+			if let Some(assignments) = self.relay_parent_assignments.remove(&old_relay_parent) {
+				let GroupAssignments { current, next } = assignments;
+
+				if let Some(cur) = current {
+					if let Entry::Occupied(mut occupied) = self.current_assignments.entry(cur) {
+						*occupied.get_mut() -= 1;
+						if *occupied.get() == 0 {
+							occupied.remove_entry();
+						}
+					}
+				}
+
+				if let Some(next) = next {
+					if let Entry::Occupied(mut occupied) = self.next_assignments.entry(next) {
+						*occupied.get_mut() -= 1;
+						if *occupied.get() == 0 {
+							occupied.remove_entry();
+						}
+					}
+				}
+			}
+		}
+	}
+
+	pub(crate) fn is_current_or_next(&self, id: ParaId) -> bool {
+		self.current_assignments.contains_key(&id) || self.next_assignments.contains_key(&id)
+	}
+}
+
+pub(crate) type CollatorFitness = lru::LruCache<CollatorId, CollatorFitnessMetric>;
+
+pub(crate) type Reservoir = HashMap<PeerId, PeerData>;
+
+pub struct PeerSlots {
+	seq_nr: u64,
+	pub(crate) fitness: CollatorFitness,
+	pub(crate) peer_data: Reservoir,
+	pub(crate) active_paras: ActiveParas,
+	stale_peers: Vec<PeerId>,
+}
+
+impl Default for PeerSlots {
+	fn default() -> Self {
+		Self {
+			seq_nr: 0u64,
+			fitness: lru::LruCache::new(CACHE_SIZE),
+			peer_data: HashMap::new(),
+			active_paras: ActiveParas::default(),
+			stale_peers: Vec::new(),
+		}
+	}
+}
+
+impl std::fmt::Debug for PeerSlots {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+		write!(f, "Eladed")
+	}
+}
+
+impl PeerSlots {
+	pub fn insert_collator(&mut self, collator_id: &CollatorId) {
+		if self.fitness.get(collator_id).is_none() {
+			self.fitness.put(collator_id.clone(), CollatorFitnessMetric::default());
+		}
+	}
+
+	pub fn stale_peers(&mut self) -> Vec<PeerId> {
+		self.stale_peers.drain(0..).collect()
+	}
+
+	pub fn reset_sample(&mut self) {
+		self.seq_nr = 0;
+	}
+}
+
+pub(crate) fn handle_view_change(peer_slots: &mut PeerSlots, view: &View) {
+	for (peer_id, peer_data) in peer_slots.peer_data.iter_mut() {
+		peer_data.prune_old_advertisements(view);
+
+		if let Some(para_id) = peer_data.collating_para() {
+			if !peer_slots.active_paras.is_current_or_next(para_id) {
+				 peer_slots.stale_peers.push(*peer_id);
+			}
+		}
+	}
+}
+
+pub(crate) async fn cycle_para(
+	sender: &mut impl SubsystemSender,
+	active_paras: &mut ActiveParas,
+	keystore: &SyncCryptoStorePtr,
+	new_relay_parents: impl IntoIterator<Item = Hash>,
+	old_relay_parents: impl IntoIterator<Item = Hash>,
+) {
+	active_paras.assign_incoming(sender, keystore, new_relay_parents).await;
+	active_paras.remove_outgoing(old_relay_parents);
+}
+
+fn peers_to_evict(peer_data: &Reservoir, fitness: &mut CollatorFitness, num: usize, rng: &mut ThreadRng) -> Vec<PeerId> {
```

Comment:
I suggest either pass a generic Rng in, or call `thread_rng` internally and remove the parameter. Generic RNG may be better for unit-testing